### PR TITLE
feat: provide plugin handling

### PR DESCRIPTION
**Pull request by didoda**
This provides cakephp plugin handle.
Example use (for a module plugin):

- create a plugin `dogs` in `plugins`
- load plugin in bedita4web `config/app.php`:
```
    'Plugins' => [
        'Dogs' => ['autoload' => true, 'bootstrap' => true, 'routes' => true],
    ],
```
In `plugins/Dogs/config/bootstrap.php`:
```
use Cake\Core\Configure;

Configure::write('Modules.plugins', (array)Configure::read('Modules.plugins') + [['name' => 'dogs']]);
```
In `plugins/Dogs/config/routes.php`:
```
use Cake\Routing\RouteBuilder;
use Cake\Routing\Router;

Router::plugin(
    'Dogs',
    ['path' => '/'],
    function (RouteBuilder $routes) {
        $routes->connect(
            '/dogs',
            ['controller' => 'Dogs', 'action' => 'index']
        );
        $routes->connect(
            '/dogs/view/:id',
            ['controller' => 'Dogs', 'action' => 'view'],
            ['pass' => ['id']]
        );
        $routes->connect(
            '/dogs/create',
            ['controller' => 'Dogs', 'action' => 'update']
        );
    }
);
```
- enjoy

### CODE REVIEW

File: config/app.default.php [hunk 1]
```diff
@@ -408,4 +408,18 @@
...
+     * Where options array may contain
+     *
+     * - `debugOnly` - boolean - (default: false) Whether or not you want to load the plugin when in 'debug' mode only
+     * - `bootstrap` - boolean - (default: false) Whether or not you want the $plugin/config/bootstrap.php file loaded.
+     * - `routes` - boolean - (default: false) Whether or not you want to load the $plugin/config/routes.php file.
+     * - `ignoreMissing` - boolean - (default: false) Set to true to ignore missing bootstrap/routes files.
+     * - `autoload` - boolean - (default: false) Whether or not you want an autoloader registered
+     */
+    'Plugins' => [
+        'Dogs' => ['autoload' => true, 'bootstrap' => true, 'routes' => true], // a simple Dogs plugin
```
Comment (stefanorosanelli) [inline config/app.default.php]: Probably a commented `MyPlugin` is more clear than an uncommented `Dogs` :smile:

File: src/Controller/Component/ModulesComponent.php [hunk 2]
```diff
@@ -119,6 +119,10 @@ public function getModules()
                 return -$idx;
             })
             ->toList();
+        $plugins = Configure::read('Modules.plugins');
+        if ($plugins) {
+            $modules = array_merge($modules, $plugins);
```
Comment (fquffio) [inline src/Controller/Component/ModulesComponent.php]: Could you please add a test case for this line?

I'm afraid you'll have to add a fourth argument to `ModulesComponentTest::testGetModules()`.

File: src/Plugin.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+use Cake\Core\Configure;
+use Cake\Core\Plugin as CakePlugin;
+
+/**
+ * {@inheritDoc}
+ *
+ * Extended with BEdita plugins utilities
+ */
+class Plugin extends CakePlugin
```
Comment (fquffio) [inline src/Plugin.php]: Could it be possible to borrow tests for this class from `bedita/core`?

**Comment by fquffio** [comment]
Just a side note:

```php
// This WILL NOT work as expected.
\Cake\Core\Configure::write(
  'Modules.plugins',
  (array)\Cake\Core\Configure::read('Modules.plugins') + [['name' => 'dogs']]
);
```

This is due to the fact that `+` merges array basing on keys only, thus `[1, 2] + [3, 4, 5] === [1, 2, 5]`.

```php
// This will do the job.
\Cake\Core\Configure::write(
  'Modules.plugins',
  array_merge(
    (array)\Cake\Core\Configure::read('Modules.plugins'),
    [['name' => 'dogs']]
  )
);
```

**Pull request merged**